Dragon stream7 bjs2 41115

Darkselect opened the pull request.

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

Darkselect committed:
Merge branch 'Dragon-master-stream7' of https://github.com/CorporationX/project_service into Dragon-stream7-BJS2-41115

# Conflicts:
#	src/main/java/faang/school/projectservice/exception/GlobalExceptionHandler.java

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(2.0 version)

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+    @PostMapping("/update/{meetId}")
+    @Operation(summary = "Update a meeting by providing the meeting ID.")
+    public ResponseEntity<MeetDto> updateMeeting(@Valid @RequestBody MeetDto updateMeetDto,
+                                                 @PathVariable @Positive long meetId) {
```
В целом это не влияет на выполнение кода, но логично сначала указывать @PathVariable, а потом @RequestBody, т.к. мы сначала идентифицируем какой-то ресурс при помощи переменной, а потом для операции по этой переменной передаем данные для дто

Darkselect replied:
Это уже мелочь

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+public class MeetingController {
+    private final MeetingService meetingService;
+
+    @PostMapping("/create")
```
Не критично и не влияет на работу, но в названиях пути лучше не использовать глаголы. Представь, что это названия папок в Винде. Ты заходишь в папку Meetings и в ней создаешь (POST) новую сущность. Ты же не идешь для этого в папку Create.

И путь этого эндпойнта можно не указывать таким образом

В лекции 8 спринта Влад этого касается.

Darkselect replied:
Влад разрешал так делать. Согласен с тобой что не критично

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+    @PostMapping("/update/{meetId}")
+    @Operation(summary = "Update a meeting by providing the meeting ID.")
+    public ResponseEntity<MeetDto> updateMeeting(@Valid @RequestBody MeetDto updateMeetDto,
+                                                 @PathVariable @Positive long meetId) {
```
Для @Positive можно добавить message на случай исключения при валидации

Darkselect replied:
добавил

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+        return ResponseEntity.status(HttpStatus.CREATED).body(meetingService.createMeeting(createMeetDto));
+    }
+
+    @PostMapping("/update/{meetId}")
```
update в пути можно опустить

Darkselect replied:
убрал

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+    @Operation(summary = "Delete a meeting, only if the user is the project owner")
+    public ResponseEntity<Void> deleteMeeting(@PathVariable @Positive Long projectId,
+                                              @PathVariable @Positive Long currentUserId,
+                                              @PathVariable @Positive long meetId) {
```
message для аннотаций валидации

Darkselect replied:
ок

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+        return ResponseEntity.ok(meetingService.updateMeeting(updateMeetDto, meetId));
+    }
+
+    @DeleteMapping("/delete/{projectId}/{currentUserId}/{meetId}")
```
тут прям мешанина в пути (хотя все будет работать). 
Во-первых, можно убрать delete
Во-вторых, если корневой путь у нас meetings, то логично работать с meetId
В третьих, можно currentUserId(а чем логически current от просто user отличается?) и projectId принимать параметрами. И в сигнатуре прописать как @RequestParam

И получится "/{meetId}"

Darkselect replied:
переделал

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+        return ResponseEntity.noContent().build();
+    }
+
+    @PostMapping("/filters")
```
Вот это прям нехорошо. Этот эндойнт не просто фильтрует, а получает встречи, что значит GET. И тогда фильтры пойдут опциональными параметрами, т.к. Get не позволяет body

Darkselect replied:
Тут опечатка сорри

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @PostMapping("/filters")
+    @Operation(summary = "Filter meetings based on provided criteria")
```
Если я правильно понимаю, то основная цель метода не Filter, а Get

Darkselect replied:
Ok

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+        return ResponseEntity.ok(meetingService.getAllMeetings());
+    }
+
+    @GetMapping("/meeting/{meetId}")
```
Можно meeting опустить

Darkselect replied:
Посчитал что так более точно

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+
+    @GetMapping("/meeting/{meetId}")
+    @Operation(summary = "Get a meeting by its ID")
+    public ResponseEntity<MeetDto> getMeetingById(@PathVariable @Positive long meetId) {
```
message

Darkselect replied:
Ok

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+
+    @PostMapping("/filters")
+    @Operation(summary = "Filter meetings based on provided criteria")
+    public ResponseEntity<List<MeetDto>> filterMeetings(@RequestBody MeetDto filter) {
```
а не лучше ли иметь некую MeetingFilterDto?

Darkselect replied:
Нет, я для меня так удобнее :)

drewbal commented on src/main/java/faang/school/projectservice/dto/meet/MeetDto.java:
```diff
@@ -0,0 +1,40 @@
+    @NotNull(message = "ID is required")
+    private long id;
+
+    @NotNull(message = "Title is required")
```
лучше @NotBlank для String

Darkselect replied:
Ok

Darkselect replied:
ок

drewbal commented on src/main/java/faang/school/projectservice/filter/meet/MeetDateFilter.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<Meet> apply(Stream<Meet> dataStream, MeetDto filter) {
+        return dataStream.filter(meet -> meet.getCreatedAt().equals(filter.getCreatedAt()));
```
мне кажется, что ты не тестировал фильтры, иначе бы понял, что этот фильтр не сработает :))
У тебя дата создания типа LocalDateTime, а значит включает в себя дату с временем до миллисекунд. И мы будем просить их от юзера для фильтра?

Darkselect replied:
Забыл сорри. Сейчас исправлю

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+    private final List<Filter<Meet, MeetDto>> meetFilters;
+
+    public MeetDto createMeeting(MeetDto createMeetDto) {
+        log.info("Trying to create meet: {}", createMeetDto);
```
этот лог не нужен. Буквально наносекунду назад ты логировал действие в контроллере, и между этими строками ничего не произошло.

Лучше логируй результат в конце метода

Darkselect replied:
Убрал. Все сделал

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+
+    public MeetDto createMeeting(MeetDto createMeetDto) {
+        log.info("Trying to create meet: {}", createMeetDto);
+        validateUserExists(createMeetDto.getId());
```
А может еще провалидировать, существует ли проект, к которому делается встреча?

Darkselect replied:
Сделал

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+                .orElseThrow(() -> {
+                    logAndThrowMeetingNotFound(meetId);
+                    return new EntityNotFoundException("Meeting not found");
+                });
```
Я бы вынес это в отдельный метод. Это чуть ли не главный метод этого сервиса, позволяющий вытащить сущность через репо из БД

drewbal replied:
Кстати, ты его создал ниже :) но забыл убрать отсюда повтор кода

Darkselect replied:
Я не могу, я тут валидацию провожу. Прочитай внимательно условие задачи

drewbal commented on src/main/java/faang/school/projectservice/validator/MeetingValidator.java:
```diff
@@ -0,0 +1,32 @@
+@Component
+@Slf4j
+public class MeetingValidator {
+    public void validateMeetingUpdate(@Valid MeetDto updateMeetDto, long meetId, Meet meeting) {
```
для чего тут @Valid, если ты уже отвалидировал в контроллере?

Darkselect replied:
убрал

drewbal commented on src/main/java/faang/school/projectservice/validator/MeetingValidator.java:
```diff
@@ -0,0 +1,32 @@
+        } else {
+            log.info("Meeting with ID {} is updated", meetId);
+            meeting.setStatus(updateMeetDto.getMeetStatus());
+            meeting.setUpdatedAt(LocalDateTime.now(ZoneId.of("UTC")));
```
вот мне не нравится, что валидатор лезет в дела сервиса и апдейтит сущность. Он должен просто проверять, ответать Да/Нет на поставленный вопрос

Darkselect replied:
Вы вообще выносите репозиторий в валидатор. Так почему я не могу сделать так? Тем более что это связано

Darkselect replied:
Странно что Руслан разрешает вам выносить репозиторий в валидатор

RuslanZinovyev replied:
Андрей прав, валидатор не должен заниматься делами сервиса, ты изменеяешь состояние объекта и это работа сервиса, репозиторий вполне может быть зависимостью валидатора, потому что валидатор обращается в базу данных при проверке существования сущности. Не вижу в этом ничего странного.  Все зависит от архитектуры твоего приложения, если ты можешь написать валидатор без обращения к базе то отлично. Такой вариант будет также хорошо, тут же ты передаешь бизнес логику валидатору, а он должен отвечать только за валидацию.

Darkselect replied:
Понял принял

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+                });
+
+        meetingValidator.validateMeetingUpdate(updateMeetDto, meetId, meeting);
+
```
А если в дто указан id проекта, которого нет в базе?

Darkselect replied:
404 not found

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    public void deleteMeeting(Long projectId, Long currentUserId, long meetId) {
+        Project project = projectService.getProjectById(projectId);
```
сначала я бы проверил, есть ли вообще встреча с таким id

Darkselect replied:
Добавил

drewbal commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -73,4 +74,11 @@ public void validateMentorPresenceInProjectTeam(InternshipCreatedDto internShipC
+
+    public void checkUserIsProjectOwner(Long currentUserId, long meetId, Project project) {
+        if (!project.getOwnerId().equals(currentUserId)) {
+            log.error("Only the project owner can delete this meeting {}", meetId);
```
логировать ошибку лучше на самом верхнем уровне в GlobalExceptionHandler

Darkselect replied:
Понял спасибо

drewbal commented on src/main/java/faang/school/projectservice/dto/meet/MeetDto.java:
```diff
@@ -0,0 +1,40 @@
+    private MeetStatus meetStatus;
+    private LocalDateTime createdAt;
+    private LocalDateTime updatedAt;
+}
```
Ну и довольно странно, что ты создаешь встречу, но не указываешь приглашенных, правда? :)

Darkselect replied:
private List<Long> participants;

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+
+    @GetMapping
+    @Operation(summary = "Get all meetings")
+    public ResponseEntity<List<MeetDto>> getAllMeetings() {
```
а может мы все же хотим мероприятия команды или проекта? а не все-все мероприятия соцсети (которые кстати будут засекречены)

Darkselect replied:
@GetMapping("/projects/{projectId}")
    @Operation(summary = "Get all meetings of a project")
    public ResponseEntity<List<MeetDto>> getAllMeetings(@PathVariable Long projectId) {
        log.info("Getting all meetings of project {}", projectId);
        ret

drewbal commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @DeleteMapping("/delete/{projectId}/{currentUserId}/{meetId}")
+    @Operation(summary = "Delete a meeting, only if the user is the project owner")
```
кстати, мне кажется, что в описании задачи явная опечатка. Отменить встречу может создавший ее пользователь, не обязательно для этого самого главного дергать

Darkselect replied:
Я не буду отходить от условий задачи. Как попросили сделать, так и сделал

RuslanZinovyev replied:
`Удаление встречи. Удалять участников из проекта может только владелец проекта.`

Насколько я понимаю задание удалять встречу вполне может создатель встречи, а вот удалять участников из проекта может только владелец проекта что логично. Я не вижу что в задании написано что только владелец проекта может удалять встречи.

drewbal commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -73,4 +74,11 @@ public void validateMentorPresenceInProjectTeam(InternshipCreatedDto internShipC
         }
     }
+
+    public void checkUserIsProjectOwner(Long currentUserId, long meetId, Project project) {
```
я думаю, что проверка на владельца не нужна, удалять встречу может создавший ее юзер с любой ролью

Darkselect replied:
Прочитай внимательно условие задачи

Darkselect replied:
Прочитай условие

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+        return filterMeet.map(meetMapper::toDto).toList();
+    }
+
+    public List<MeetDto> getAllMeetings() {
```
надо изменить метод на получение встреч своего проекта

Darkselect replied:
@GetMapping("/projects/{projectId}")
    @Operation(summary = "Get all meetings of a project")
    public ResponseEntity<List<MeetDto>> getAllMeetings(@PathVariable Long projectId) {
        log.info("Getting all meetings of project {}", projectId);
        return ResponseEntity.ok(meetingService.getAllMeetings(projectId));
    }

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+        userServiceClient.getUser(userId);
+    }
+
+    private void logAndThrowMeetingNotFound(long meetId) {
```
тут это не нужно. Логируй ошибки в exception handler

Darkselect replied:
убрал

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+    public MeetDto getMeetingById(long meetId) {
+        return meetRepository.findById(meetId).map(meetMapper::toDto)
+                .orElseThrow(() -> {
+                    logAndThrowMeetingNotFound(meetId);
```
логируй в ExceptionHandler

Darkselect replied:
ок

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+    public MeetDto updateMeeting(MeetDto updateMeetDto, long meetId) {
+        Meet meeting = meetRepository.findById(meetId)
+                .orElseThrow(() -> {
+                    logAndThrowMeetingNotFound(meetId);
```
логируй в ExceptionHandler

Darkselect replied:
ок

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+        return meet;
+    }
+
+    private void validateUserExists(long userId) {
```
Может лучше валидацию через валидатор подумать?

drewbal replied:
И для чего лезть в другой сервис, если можно проверить наличие юзера через TeamMember

drewbal commented on src/main/java/faang/school/projectservice/validator/MeetingValidator.java:
```diff
@@ -0,0 +1,32 @@
+public class MeetingValidator {
+    public void validateMeetingUpdate(@Valid MeetDto updateMeetDto, long meetId, Meet meeting) {
+        if (meeting.getStatus().equals(MeetStatus.CANCELLED) && updateMeetDto.getMeetStatus().equals(MeetStatus.CANCELLED)) {
+            log.error("Meeting with ID {} is already cancelled", meetId);
```
логи в хэндлер

Darkselect replied:
ок

drewbal commented on src/main/java/faang/school/projectservice/validator/MeetingValidator.java:
```diff
@@ -0,0 +1,32 @@
+        }
+
+        if (!(meeting.getCreatorId() == updateMeetDto.getCreatorId())) {
+            log.error("Unauthorized access to meet with ID {}", meetId);
```
логи в хэндлер

Darkselect replied:
ок

drewbal commented on src/test/java/faang/school/projectservice/controller/MeetingControllerTest.java:
```diff
@@ -0,0 +1,141 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@WebMvcTest
```
Расскажешь про этот подход?

Darkselect replied:
Если интересно можем пообщаться в пп

drewbal commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,100 @@
+    public MeetDto createMeeting(MeetDto createMeetDto) {
+        log.info("Trying to create meet: {}", createMeetDto);
+        validateUserExists(createMeetDto.getId());
+
```
И по идее нам надо проверить существование приглашенных юзеров

Darkselect replied:
Сделал

drewbal commented on src/test/java/faang/school/projectservice/service/MeetingServiceTest.java:
```diff
@@ -0,0 +1,226 @@
+
+        verify(userServiceClient).getUser(1L);
+        verify(meetRepository, never()).save(any());
+    }
```
А не будем тестировать что произойдет, если кривой id проекта?

drewbal commented on src/test/java/faang/school/projectservice/service/MeetingServiceTest.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    @Test
+    void testDeleteMeetingWhenUserIsProjectOwner() {
```
подумай над логикой владельца проекта или встречи

Darkselect replied:
Сделал

drewbal commented on src/test/java/faang/school/projectservice/service/MeetingServiceTest.java:
```diff
@@ -0,0 +1,226 @@
+        project.setOwnerId(ownerId);
+        return project;
+    }
+}
```
Почему ты не тестируешь, работают ли твои фильтры?

Darkselect replied:
Забыл

drewbal commented on src/test/java/faang/school/projectservice/validator/MeetingValidatorTest.java:
```diff
@@ -0,0 +1,86 @@
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
зачем, если ты делаешь @ExtendWith?

Darkselect replied:
понял

RuslanZinovyev replied:
Ты удалил MockitoAnnotations.openMocks(this);?

drewbal commented on src/test/java/faang/school/projectservice/validator/MeetingValidatorTest.java:
```diff
@@ -0,0 +1,86 @@
+        verify(meeting, times(1)).setStatus(MeetStatus.COMPLETED);
+        verify(meeting, times(1)).setUpdatedAt(any(LocalDateTime.class));
+    }
+}
```
Еще ты забыл протестить ProjectValidator

Darkselect replied:
добавил тесты туда тоже

drewbal reviewed (commented):
Дониёр, клево! Есть вопросы по логике некоторых решений (мне кажется в постановке задачи явные опечатки), ну и несколько замечаний по коду.

Тест контроллера выглядит очень круто, не знал, что так можно. Буду учиться на твоем примере

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(3.0 version)

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(4.0 version)

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(5.0 version)

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+        return ResponseEntity.status(HttpStatus.CREATED).body(meetingService.createMeeting(createMeetDto));
+    }
+
+    @PostMapping("/{meetId}")
```
PutMapping если обновляем данные

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+    public ResponseEntity<Void> deleteMeeting(
+            @PathVariable @Positive(message = "Project ID must be a positive number") Long projectId,
+            @PathVariable @Positive(message = "Meeting ID must be a positive number") long meetId,
+            @RequestBody MeetDto deleteDto) {
```
А зачем мы передаем `MeetDto `в качестве тела запроса, для чего нам это нужно? Разве мы не может удалить встречу по `meetId`?

Darkselect replied:
validateMeetCreator(deleteDto); Я смотрю это создатель встречи или нет.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+    @GetMapping("/filters")
+    @Operation(summary = "Get meetings based on provided criteria")
+    @RequestMapping("/filters")
+    public ResponseEntity<List<MeetDto>> filterMeetings(@RequestBody MeetDto meetDto) {
```
Лучше не передавать тело запроса для GET ендпоинта. Удали аннотацию @RequestBody, таким образом все поля класса MeetDto будут использованы как параметры запроса и станут опциональными. Также можешь прикрутить пагинацию если хочется сделать прям по уму.

Darkselect replied:
@GetMapping("/filters")
@Operation(summary = "Get meetings based on provided criteria")
@RequestMapping("/filters")
public ResponseEntity<List<MeetDto>> filterMeetings(MeetDto meetDto,
@PageableDefault(size = 10, page = 0) Pageable pageable) {
log.info("Filtering meetings based on provided criteria {} ", meetDto);
return ResponseEntity.ok(meetingService.filterMeetings(meetDto, pageable));
}

Darkselect replied:
ок

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+        return ResponseEntity.ok(meetingService.getAllMeetings(projectId));
+    }
+
+    @GetMapping("/meeting/{meetId}")
```
поменяй на множественное число `meetings`

Darkselect replied:
поменял

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+        return ResponseEntity.ok(meetingService.filterMeetings(meetDto));
+    }
+
+    @GetMapping("/projects/{projectId}")
```
В URL ожидается проект, а сам метод вытаскивает все митинги по id

Darkselect replied:
исправил. Посмотришь пожалуйста

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+        return ResponseEntity.ok(meetingService.updateMeeting(updateMeetDto, meetId));
+    }
+
+    @DeleteMapping("/{projectId}/{meetId}")
```
Переделай в `/projects/{projectId}/meetings/{meetingId}`

Darkselect replied:
ok

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+    public ResponseEntity<MeetDto> getMeetingById(
+            @PathVariable
+            @Positive(message = "Meeting ID must be a positive number")
+            long meetId
```
поменяй на Long

Darkselect replied:
изменил. Согласен да там может ничего и не быть

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+    @Operation(summary = "Delete a meeting, only if the user is the project owner")
+    public ResponseEntity<Void> deleteMeeting(
+            @PathVariable @Positive(message = "Project ID must be a positive number") Long projectId,
+            @PathVariable @Positive(message = "Meeting ID must be a positive number") long meetId,
```
Поменяй на Long

Darkselect replied:
ok

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -0,0 +1,86 @@
+    @Operation(summary = "Update a meeting by providing the meeting ID.")
+    public ResponseEntity<MeetDto> updateMeeting(
+            @PathVariable
+            @Positive(message = "Meeting ID must be a positive number") long meetId,
```
Поменяй на Long

Darkselect replied:
ок

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/meet/MeetDto.java:
```diff
@@ -0,0 +1,43 @@
+public class MeetDto {
+    @Positive(message = "ID must be a positive number")
+    @NotNull(message = "ID is required")
+    private long id;
```
Примитивный тип не может быть NULL, если хочешь валидировать на NULL то измени тип на Long

Darkselect replied:
Тут опечатка сорри. Подправил

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/MeetingService.java:
```diff
@@ -0,0 +1,99 @@
+
+        projectValidator.validateProjectExistsById(updateMeetDto.getProjectId());
+
+        meetingValidator.validateMeetingUpdate(updateMeetDto, meetId, meeting);
```
Нам не нужен метод `validateMeetingUpdate`, валидтор должен заниматься валидацией, а не изменением сущности и сохранением ее в базе, перенеси всю эту логику сюда в сервис. Так будет более логично и правильно.

Darkselect replied:
Ок сделал его приватным 

private void validateMeetingUpdate(MeetDto updateMeetDto, long meetId, Meet meeting) {
if (meeting.getStatus().equals(MeetStatus.CANCELLED) && updateMeetDto.getMeetStatus().equals(MeetStatus.CANCELLED)) {
throw new MeetingAlreadyCancelledException("Meeting is already cancelled");
} else {
log.info("Meeting with ID {} is updated", meetId);
meeting.setStatus(updateMeetDto.getMeetStatus());
meeting.setUpdatedAt(LocalDateTime.now(ZoneId.of("UTC")));
}

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(6.0 version)

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -37,48 +39,47 @@ public ResponseEntity<MeetDto> createMeeting(@Valid @RequestBody MeetDto createM
-            @PathVariable @Positive(message = "Project ID must be a positive number") Long projectId,
-            @PathVariable @Positive(message = "Meeting ID must be a positive number") long meetId,
+            @PathVariable @Positive(message = "Meeting ID must be a positive number") Long meetId,
             @RequestBody MeetDto deleteDto) {
```
Передай creatorId и удали MeetDto

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(7.0 version)

Darkselect committed:
Merge branch 'Dragon-master-stream7' of https://github.com/CorporationX/project_service into Dragon-stream7-BJS2-41115

# Conflicts:
#	src/test/java/faang/school/projectservice/validator/ProjectValidatorTest.java

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(7.0 version)

Darkselect committed:
Merge branch 'Dragon-master-stream7' of https://github.com/CorporationX/project_service into Dragon-stream7-BJS2-41115

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/InternshipController.java
#	src/main/java/faang/school/projectservice/exception/GlobalExceptionHandler.java

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(8.0 version)

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/MeetingController.java:
```diff
@@ -49,13 +49,13 @@ public ResponseEntity<MeetDto> updateMeeting(
     }
 
-    @DeleteMapping("/meetings/{meetId}")
+    @DeleteMapping("/meetings/{meetId}/{creatorId}")
```
Мы не кладем один Id за другим Id без пояcнения, `/meetings/{meetId}` - тут понятно к чему относится id
`{creatorId}` - а вот тут не понятно, тогда надо сделать `/meetings/{meetId}/creators/{creatorId}`

Darkselect committed:
BJS2-41115 Task

Implement meeting management functionality: create, update, delete, and filter meetings

- Added functionality for creating meetings with validation of project and user existence.

- Implemented meeting update logic, ensuring only the creator can update the meeting and handling status changes, including cancellation.

- Introduced deletion logic, allowing only the project owner to delete meetings.

- Added timestamp management for meeting creation and last update.

- Integrated filtering functionality for meetings by date and title.

- Provided methods to retrieve all meetings, meetings by ID, and meeting filters.

- Enhanced error handling for missing entities (meetings, users, projects).

- Refactored the meeting initialization process to set default values (status, timestamps).

With All Tests

(9.0 version)
